src/txHashBuilder.c

```diff
@@ -288,20 +291,352 @@ void txHashBuilder_addCertificate_delegation(
+	builder->poolCertificateData.remainingRelays--;
+
+	ASSERT(dnsName != NULL);
+	ASSERT(dnsNameSize > 0);
+	ASSERT(dnsNameSize < BUFFER_SIZE_PARANOIA);
+
+	// Array(3)[
+	//   Unsigned[1]
+	//   Unsigned[port] / Null
+	//   Text[dnsName]
+	// ]
+	{
+		BUILDER_APPEND_CBOR(CBOR_TYPE_ARRAY, 3);
+		{
+			BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, 1);
+		}
+		{
+			if (port != NULL) {
+				BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, *port);
+			} else {
+				BUILDER_APPEND_CBOR(CBOR_TYPE_NULL, 0);
+			}
+		}
+		{
+			BUILDER_APPEND_CBOR(CBOR_TYPE_TEXT, dnsNameSize);
+			BUILDER_APPEND_DATA(dnsName, dnsNameSize);
+		}
+	}
+}
+
+void txHashBuilder_addPoolRegistrationCertificate_addRelay2(
+        tx_hash_builder_t* builder,
+        const uint8_t* dnsName, size_t dnsNameSize
+)
+{
+	TRACE("txHashBuilder_addPoolRegistrationCertificate_addRelay2: %d", builder->state);
+
+	addRelay_updateState(builder);
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL_RELAYS);
+	ASSERT(builder->poolCertificateData.remainingRelays > 0);
+	builder->poolCertificateData.remainingRelays--;
+
+	ASSERT(dnsName != NULL);
+	ASSERT(dnsNameSize > 0);
+	ASSERT(dnsNameSize < BUFFER_SIZE_PARANOIA);
+
+	// Array(2)[
+	//   Unsigned[2]
+	//   Text[dnsName]
+	// ]
+	{
+		BUILDER_APPEND_CBOR(CBOR_TYPE_ARRAY, 3);
+		{
+			BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, 1);
+		}
+		{
+			BUILDER_APPEND_CBOR(CBOR_TYPE_TEXT, dnsNameSize);
+			BUILDER_APPEND_DATA(dnsName, dnsNameSize);
+		}
+	}
+}
+
+// enter empty owners or relays if none were received
+static void addPoolMetadata_updateState(tx_hash_builder_t* builder)
+{
+	switch (builder->state) {
+	case TX_HASH_BUILDER_IN_CERTIFICATES_POOL:
+	case TX_HASH_BUILDER_IN_CERTIFICATES_POOL_OWNERS:
+		txHashBuilder_addPoolRegistrationCertificate_enterRelays(builder);
+	// intentional fallthrough
+
+	case TX_HASH_BUILDER_IN_CERTIFICATES_POOL_RELAYS:
+		ASSERT(builder->poolCertificateData.remainingRelays == 0);
+		break; // we want to be here
+
+	default:
+		ASSERT(false);
+	}
+
+	builder->state = TX_HASH_BUILDER_IN_CERTIFICATES_POOL_METADATA;
+}
+
+void txHashBuilder_addPoolRegistrationCertificate_addPoolMetadata(
+        tx_hash_builder_t* builder,
+        const uint8_t* url, size_t urlSize,
+        const uint8_t* metadataHash, size_t metadataHashSize
+)
+{
+	TRACE("txHashBuilder_addPoolRegistrationCertificate_addPoolMetadata: %d", builder->state);
+
+	addPoolMetadata_updateState(builder);
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL_METADATA);
+
+	// Array(2)[
+	//   Tstr[url]
+	//   Bytes[metadataHash]
+	// ]
+	{
+		BUILDER_APPEND_CBOR(CBOR_TYPE_ARRAY, 2);
+		{
+			BUILDER_APPEND_CBOR(CBOR_TYPE_TEXT, urlSize);
+			BUILDER_APPEND_DATA(url, urlSize);
+		}
+		{
+			BUILDER_APPEND_CBOR(CBOR_TYPE_BYTES, metadataHashSize);
+			BUILDER_APPEND_DATA(metadataHash, metadataHashSize);
+		}
+	}
+	builder->state = TX_HASH_BUILDER_IN_CERTIFICATES;
+}
+
+void txHashBuilder_addPoolRegistrationCertificate_addPoolMetadata_null(
```

Comment:
Pools should have metadata, I don't see why they wouldn't as that's the info by which they are searchable in online tools like pooltool: https://pooltool.io/pool/04c60c78417132a195cbb74975346462410f72612952a7c4ade7e438/blocks

Reply:
I just followed the CDDL spec which allows null, but perhaps we can ask IOHK if we should always require metadata?

Reply:
So I asked if we should support pools with no metadata, and this is what they say: 
"Yes!!! Most of our pools will be private and not have any metadata."

Reply:
Thanks for asking and sorry for the confusion, I will update Trezor implementation accordingly